File: app/src/main/java/closer/vlllage/com/closer/handler/group/GroupToolbarHandler.java
package closer.vlllage.com.closer.handler.group;

import android.support.annotation.DrawableRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import com.google.android.gms.maps.model.LatLng;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import closer.vlllage.com.closer.R;
import closer.vlllage.com.closer.handler.data.ApiHandler;
import closer.vlllage.com.closer.handler.data.PersistenceHandler;
import closer.vlllage.com.closer.handler.data.RefreshHandler;
import closer.vlllage.com.closer.handler.event.EventHandler;
import closer.vlllage.com.closer.handler.helpers.ActivityHandler;
import closer.vlllage.com.closer.handler.helpers.AlertHandler;
import closer.vlllage.com.closer.handler.helpers.DefaultAlerts;
import closer.vlllage.com.closer.handler.helpers.DisposableHandler;
import closer.vlllage.com.closer.handler.helpers.OutboundHandler;
import closer.vlllage.com.closer.handler.helpers.ResourcesHandler;
import closer.vlllage.com.closer.handler.map.MapActivityHandler;
import closer.vlllage.com.closer.pool.PoolMember;
import closer.vlllage.com.closer.store.models.Event;
import closer.vlllage.com.closer.store.models.Group;
import closer.vlllage.com.closer.store.models.Phone;
import closer.vlllage.com.closer.ui.CircularRevealActivity;
import io.reactivex.subjects.BehaviorSubject;

import static com.google.android.gms.common.util.Strings.isEmptyOrWhitespace;

public class GroupToolbarHandler extends PoolMember {

    private RecyclerView recyclerView;
    private ToolbarAdapter adapter;
    private final BehaviorSubject<Boolean> isShareActiveObservable = BehaviorSubject.createDefault(false);

    private Group group;

    public void attach(RecyclerView recyclerView) {
        this.recyclerView = recyclerView;
        adapter = new ToolbarAdapter(this);

        recyclerView.setLayoutManager(new LinearLayoutManager(recyclerView.getContext(), LinearLayoutManager.HORIZONTAL, false) {
            @Override
            public void measureChild(@NonNull View child, int widthUsed, int heightUsed) {
                child.getLayoutParams().width = recyclerView.getMeasuredWidth() / adapter.getItemCount();
                super.measureChild(child, widthUsed, heightUsed);
            }

            @Override
            public void measureChildWithMargins(@NonNull View child, int widthUsed, int heightUsed) {
                child.getLayoutParams().width = recyclerView.getMeasuredWidth() / adapter.getItemCount();
                super.measureChildWithMargins(child, widthUsed, heightUsed);
            }
        });
        recyclerView.setAdapter(adapter);

        $(DisposableHandler.class).add(isShareActiveObservable.subscribe(isShareActive -> show(group)));
        $(DisposableHandler.class).add($(GroupHandler.class).onGroupUpdated().subscribe(this::show));
        $(DisposableHandler.class).add($(GroupHandler.class).onGroupChanged().subscribe(this::show));
        $(DisposableHandler.class).add($(GroupHandler.class).onEventChanged().subscribe(g -> show(group)));
        $(DisposableHandler.class).add($(GroupHandler.class).onPhoneChanged().subscribe(g -> show(group)));
    }

    private void show(Group group) {
        if (group == null) {
            return;
        }

        this.group = group;
        Event event = $(GroupHandler.class).onEventChanged().getValue();
        Phone phone = $(GroupHandler.class).onPhoneChanged().getValue();

        List<ToolbarItem> items = new ArrayList<>();

        if (group.hasPhone()) {
            items.add(new ToolbarItem(
                    R.string.messages,
                    R.drawable.ic_message_black_24dp,
                    v -> {}
            ));

            items.add(new ToolbarItem(
                    R.string.photos,
                    R.drawable.ic_photo_black_24dp,
                    v -> {}
            ));

            items.add(new ToolbarItem(
                    R.string.groups,
                    R.drawable.ic_person_black_24dp,
                    v -> {}
            ));
        }

        if (items.size() < 3 && event != null) {
            items.add(new ToolbarItem(
                    isShareActiveObservable.getValue() ? R.string.cancel : R.string.share,
                    isShareActiveObservable.getValue() ? R.drawable.ic_close_black_24dp : R.drawable.ic_share_black_24dp,
                    v -> toggleShare()
            ));
        }

        if (items.size() < 3 && (event != null || group.isPhysical())) {
            items.add(new ToolbarItem(
                    R.string.show_on_map,
                    R.drawable.ic_my_location_black_24dp,
                    v -> {
                        if (event != null) showEventOnMap(event);
                        else showGroupOnMap(group);
                    }
            ));
        }

        if (items.size() < 3 && isEventCancelable(event)) {
            items.add(new ToolbarItem(
                    R.string.cancel,
                    R.drawable.ic_close_black_24dp,
                    v -> $(AlertHandler.class).make()
                            .setTitle($(ResourcesHandler.class).getResources().getString(R.string.cancel_event))
                            .setMessage($(ResourcesHandler.class).getResources().getString(R.string.event_will_be_cancelled, event.getName()))
                            .setPositiveButton($(ResourcesHandler.class).getResources().getString(R.string.cancel_event))
                            .setPositiveButtonCallback(result -> {
                                $(DisposableHandler.class).add($(ApiHandler.class).cancelEvent(event.getId()).subscribe(successResult -> {
                                    if (successResult.success) {
                                        $(DefaultAlerts.class).message($(ResourcesHandler.class).getResources().getString(R.string.event_cancelled, event.getName()));
                                        $(RefreshHandler.class).refreshEvents(new LatLng(event.getLatitude(), event.getLongitude()));
                                    } else {
                                        $(DefaultAlerts.class).thatDidntWork();
                                    }
                                }, error -> $(DefaultAlerts.class).thatDidntWork()));
                            })
                            .show()
            ));
        }

        if (items.size() < 3 && group.isPhysical() && isEmptyOrWhitespace(group.getName())) {
            items.add(new ToolbarItem(
                    R.string.set_name,
                    R.drawable.ic_edit_location_black_24dp,
                    v -> $(PhysicalGroupUpgradeHandler.class).convertToHub(group, updatedGroup -> {
                        $(GroupHandler.class).showGroupName(updatedGroup);
                        show(updatedGroup);
                    })
            ));
        }

        if (items.size() < 3 && group.isPhysical() && isEmptyOrWhitespace(group.getPhoto())) {
            items.add(new ToolbarItem(
                    R.string.set_background,
                    R.drawable.ic_camera_black_24dp,
                    v -> $(PhysicalGroupUpgradeHandler.class).setBackground(group, updateGroup -> {
                        $(GroupHandler.class).setGroupBackground(updateGroup);
                        show(updateGroup);
                    })
            ));
        }

        if (items.size() < 3 && (event != null || group.isPhysical())) {
            items.add(new ToolbarItem(
                    R.string.get_directions,
                    R.drawable.ic_directions_black_24dp,
                    v -> $(OutboundHandler.class).openDirections(new LatLng(
                            group.getLatitude(),
                            group.getLongitude()
                    ))
            ));
        }

        if (items.size() < 3 && group.isPhysical()) {
            items.add(new ToolbarItem(
                    R.string.host_event,
                    R.drawable.ic_event_note_black_24dp,
                    v -> $(EventHandler.class).createNewEvent(new LatLng(
                            group.getLatitude(),
                            group.getLongitude()
                    ), group.isPublic(), this::showEventOnMap)
            ));
        }

        adapter.setItems(items);
    }

    private boolean isEventCancelable(@Nullable Event event) {
        return event != null && $(PersistenceHandler.class).getPhoneId() != null &&
                !event.isCancelled() && event.getCreator() != null &&
                new Date().before(event.getEndsAt()) &&
                event.getCreator().equals($(PersistenceHandler.class).getPhoneId());
    }

    private void showGroupOnMap(Group group) {
        ((CircularRevealActivity) $(ActivityHandler.class).getActivity())
                .finish(() -> $(MapActivityHandler.class).showGroupOnMap(group));
    }

    private void toggleShare() {
        if (isShareActiveObservable.getValue()) {
            isShareActiveObservable.onNext(false);
            return;
        }

        isShareActiveObservable.onNext(true);
    }



    private void showEventOnMap(Event event) {
        ((CircularRevealActivity) $(ActivityHandler.class).getActivity())
                .finish(() -> $(MapActivityHandler.class).showEventOnMap(event));
    }

    public BehaviorSubject<Boolean> getIsShareActiveObservable() {
        return isShareActiveObservable;
    }

    static class ToolbarItem {
        View.OnClickListener onClickListener;
        @StringRes int name;
        @DrawableRes int icon;

        public ToolbarItem(int name, int icon, View.OnClickListener onClickListener) {
            this.onClickListener = onClickListener;
            this.name = name;
            this.icon = icon;
        }
    }
}

File: app/src/main/java/closer/vlllage/com/closer/handler/group/GroupToolbarHandler.kt
package closer.vlllage.com.closer.handler.group

import android.view.View
import androidx.annotation.ColorRes
import androidx.annotation.DrawableRes
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import closer.vlllage.com.closer.ContentViewType
import closer.vlllage.com.closer.R
import closer.vlllage.com.closer.handler.data.ApiHandler
import closer.vlllage.com.closer.handler.data.DataHandler
import closer.vlllage.com.closer.handler.data.PersistenceHandler
import closer.vlllage.com.closer.handler.data.RefreshHandler
import closer.vlllage.com.closer.handler.event.EventHandler
import closer.vlllage.com.closer.handler.helpers.*
import closer.vlllage.com.closer.handler.map.MapActivityHandler
import closer.vlllage.com.closer.handler.share.ShareActivityTransitionHandler
import closer.vlllage.com.closer.store.models.Event
import closer.vlllage.com.closer.store.models.Group
import closer.vlllage.com.closer.ui.CircularRevealActivity
import com.google.android.gms.maps.model.LatLng
import com.queatz.on.On
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.subjects.BehaviorSubject
import java.util.*
import java.util.concurrent.TimeUnit

class GroupToolbarHandler constructor(private val on: On) {

    private lateinit var recyclerView: RecyclerView
    private lateinit var adapter: ToolbarAdapter
    val contentView = BehaviorSubject.createDefault(ContentViewType.MESSAGES)

    fun attach(recyclerView: RecyclerView, onToolbarItemSelected: (ToolbarItem) -> Unit) {
        this.recyclerView = recyclerView
        adapter = ToolbarAdapter(on, onToolbarItemSelected)

        recyclerView.layoutManager = LinearLayoutManager(recyclerView.context, RecyclerView.HORIZONTAL, false)
        recyclerView.adapter = adapter

        on<DisposableHandler>().add(contentView.subscribe {
            show(on<GroupHandler>().group)
            adapter.selectedContentView.onNext(it)
        })

        adapter.selectedContentView.distinctUntilChanged()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe {
            scrollToolbarTo(it)
        }.also {
            on<DisposableHandler>().add(it)
        }

        on<GroupHandler>().onGroupUpdated { show(it) }
        on<GroupHandler>().onGroupChanged { show(it) }
        on<GroupHandler>().onEventChanged { show(on<GroupHandler>().group) }
        on<GroupHandler>().onPhoneChanged { show(on<GroupHandler>().group) }
    }

    private fun scrollToolbarTo(content: ContentViewType) {
        adapter.items.indexOfFirst { it.value == content }.takeIf { it >= 0 }?.let { index ->
            recyclerView.smoothScrollToPosition(index)
        }
    }

    private fun show(group: Group?) {
        if (group == null) {
            return
        }

        val event = if (group.hasEvent()) on<GroupHandler>().event else null

        val items = mutableListOf<ToolbarItem>()

        if (group.hasPhone()) {
            items.add(ToolbarItem(
                    on<ResourcesHandler>().resources.getString(R.string.about),
                    R.drawable.ic_person_black_24dp,
                    View.OnClickListener {
                        contentView.onNext(ContentViewType.PHONE_ABOUT)
                    },
                    ContentViewType.PHONE_ABOUT
            ))

            items.add(ToolbarItem(
                    on<ResourcesHandler>().resources.getString(R.string.posts),
                    R.drawable.ic_view_day_black_24dp,
                    View.OnClickListener {
                        contentView.onNext(ContentViewType.PHONE_MESSAGES)
                    },
                    ContentViewType.PHONE_MESSAGES
            ))

            items.add(ToolbarItem(
                    on<ResourcesHandler>().resources.getString(R.string.photos),
                    R.drawable.ic_photo_black_24dp,
                    View.OnClickListener {
                        contentView.onNext(ContentViewType.PHONE_PHOTOS)
                    },
                    ContentViewType.PHONE_PHOTOS
            ))

            items.add(ToolbarItem(
                    on<ResourcesHandler>().resources.getString(R.string.groups),
                    R.drawable.ic_group_black_24dp,
                    View.OnClickListener {
                        contentView.onNext(ContentViewType.PHONE_GROUPS)
                    },
                    ContentViewType.PHONE_GROUPS
            ))

            items.add(ToolbarItem(
                    on<ResourcesHandler>().resources.getString(R.string.invite),
                    R.drawable.ic_person_add_black_24dp,
                    View.OnClickListener {
                        on<ShareActivityTransitionHandler>().inviteToGroup(group.phoneId!!)
                    }
            ))

            items.add(ToolbarItem(
                    on<ResourcesHandler>().resources.getString(R.string.show_on_map),
                    R.drawable.ic_my_location_black_24dp,
                    View.OnClickListener {
                        val runnable = { on<MapActivityHandler>().showPhoneOnMap(group.phoneId) }
                        if (on<ActivityHandler>().activity is CircularRevealActivity) {
                            (on<ActivityHandler>().activity as CircularRevealActivity).finish(runnable)
                        } else {
                            runnable.invoke()
                        }
                    }
            ))

            items.add(ToolbarItem(
                    on<ResourcesHandler>().resources.getString(R.string.get_directions),
                    R.drawable.ic_directions_black_24dp,
                    View.OnClickListener {
                        on<DataHandler>().getPhone(group.phoneId!!).subscribe({
                            if (it.latitude != null && it.longitude != null) {
                                on<OutboundHandler>().openDirections(LatLng(it.latitude!!, it.longitude!!
                                ))
                            } else {
                                on<DefaultAlerts>().thatDidntWork()
                            }
                        }, {
                            on<DefaultAlerts>().thatDidntWork()
                        })
                    }
            ))
        }

        if (group.physical || (!group.hasEvent() && !group.isPublic)) {
            items.add(ToolbarItem(
                    on<ResourcesHandler>().resources.getString(R.string.events),
                    R.drawable.ic_event_available_black_24dp,
                    View.OnClickListener {
                        contentView.onNext(ContentViewType.EVENTS)
                    },
                    ContentViewType.EVENTS
            ))
        }

        if (group.ratingCount ?: 0 > 0 && (group.physical || group.hasEvent())) {
            items.add(ToolbarItem(
                    on<ResourcesHandler>().resources.getQuantityString(R.plurals.review_count, group.ratingCount!!, group.ratingCount!!),
                    R.drawable.ic_star_black_24dp,
                    View.OnClickListener {
                        contentView.onNext(ContentViewType.REVIEWS)
                    },
                    ContentViewType.REVIEWS
            ))
        }

        if (!group.hasPhone()) {
            items.add(ToolbarItem(
                    on<ResourcesHandler>().resources.getString(R.string.members),
                    R.drawable.ic_group_black_24dp,
                    View.OnClickListener {
                        contentView.onNext(ContentViewType.CONTACTS)
                    },
                    ContentViewType.CONTACTS
            ))
        }

        if (!group.hasPhone() && group.isPublic && on<GroupMemberHandler>().isCurrentUserMemberOf(group).not()) {
            items.add(ToolbarItem(
                    on<ResourcesHandler>().resources.getString(if (group.hasEvent()) R.string.join_event else R.string.join_group),
                    R.drawable.ic_person_add_black_24dp,
                    View.OnClickListener {
                        on<GroupMemberHandler>().join(group)
                    }
            ))
        }

        if (isEventEnded(event)) {
            items.add(ToolbarItem(
                    on<ResourcesHandler>().resources.getString(R.string.host_again),
                    R.drawable.ic_refresh_black_24dp,
                    View.OnClickListener {
                        event?.let {
                            val daysAgo = TimeUnit.MILLISECONDS.toDays(Date().time - event.startsAt!!.time).toInt()
                            val now = Calendar.getInstance(TimeZone.getDefault())
                            val startsAt = event.startsAt!!.let { Calendar.getInstance(TimeZone.getDefault()).apply {
                                time = it
                                add(Calendar.DATE, daysAgo)
                            } }
                            val endsAt = event.endsAt!!.let { Calendar.getInstance(TimeZone.getDefault()).apply {
                                time = it
                                add(Calendar.DATE, daysAgo)
                            } }

                            if (startsAt.before(now)) {
                                startsAt.add(Calendar.DATE, 1)
                                endsAt.add(Calendar.DATE, 1)
                            }

                            on<EventHandler>().createNewEvent(
                                    LatLng(it.latitude!!, it.longitude!!),
                                    it.isPublic,
                                    event.name,
                                    event.about,
                                    startsAt.time,
                                    endsAt.time
                            ) {
                                on<GroupActivityTransitionHandler>().showGroupForEvent(null, it)
                            }
                        }
                    }
            ))
        }

        if (event != null) {
            items.add(ToolbarItem(
                    on<ResourcesHandler>().resources.getString(R.string.share),
                    R.drawable.ic_share_black_24dp,
                    View.OnClickListener { shareEvent(event) }
            ))
        }

        if (event != null || group.physical) {
            items.add(ToolbarItem(
                    on<ResourcesHandler>().resources.getString(R.string.show_on_map),
                    R.drawable.ic_my_location_black_24dp,
                    View.OnClickListener { v ->
                        if (event != null)
                            showEventOnMap(event)
                        else
                            showGroupOnMap(group)
                    }
            ))
        }

        if (group.physical && group.name.isNullOrBlank()) {
            items.add(ToolbarItem(
                    on<ResourcesHandler>().resources.getString(R.string.set_name),
                    R.drawable.ic_edit_location_black_24dp,
                    View.OnClickListener { v ->
                        on<PhysicalGroupUpgradeHandler>().convertToHub(group) { updatedGroup ->
                            on<RefreshHandler>().refresh(updatedGroup)
                            show(updatedGroup)
                        }
                    }
            ))
        }

        if (group.physical && group.photo.isNullOrBlank()) {
            items.add(ToolbarItem(
                    on<ResourcesHandler>().resources.getString(R.string.set_background),
                    R.drawable.ic_camera_black_24dp,
                    View.OnClickListener { v ->
                        on<PhysicalGroupUpgradeHandler>().setBackground(group) { updatedGroup ->
                            on<RefreshHandler>().refresh(updatedGroup)
                            show(updatedGroup)
                        }
                    }
            ))
        }

        if (!group.hasPhone() && !group.hasEvent()) {
            items.add(ToolbarItem(
                    on<ResourcesHandler>().resources.getString(R.string.add_action),
                    R.drawable.ic_add_black_24dp,
                    View.OnClickListener {
                        on<GroupActionHandler>().addActionToGroup(group)
                    }
            ))
        }

        if ((event != null || group.physical)) {
            items.add(ToolbarItem(
                    on<ResourcesHandler>().resources.getString(R.string.get_directions),
                    R.drawable.ic_directions_black_24dp,
                    View.OnClickListener { v ->
                        on<OutboundHandler>().openDirections(LatLng(
                                group.latitude!!,
                                group.longitude!!
                        ))
                    }
            ))
        }

        if (group.physical || (!group.hasEvent() && !group.isPublic)) {
            items.add(ToolbarItem(
                    on<ResourcesHandler>().resources.getString(R.string.host_event),
                    R.drawable.ic_event_note_black_24dp,
                    View.OnClickListener {
                        on<HostEventHelper>().hostEvent(group)
                    }
            ))
        }

        if (group.physical || group.hasEvent()) {
            items.add(ToolbarItem(
                    on<ResourcesHandler>().resources.getString(R.string.write_review),
                    R.drawable.ic_star_half_black_24dp,
                    View.OnClickListener {
                        on<ReviewHandler>().postReview(group)
                    }
            ))
        }

        if (!group.hasPhone() && items.isNotEmpty()) {
            items.add(0, ToolbarItem(
                    on<ResourcesHandler>().resources.getString(R.string.posts),
                    R.drawable.ic_view_day_black_24dp,
                    View.OnClickListener {
                        contentView.onNext(ContentViewType.MESSAGES)
                    },
                    ContentViewType.MESSAGES
            ))
        }

        if (isEventCancelable(event)) {
            items.add(ToolbarItem(
                    on<ResourcesHandler>().resources.getString(R.string.cancel),
                    R.drawable.ic_close_black_24dp,
                    View.OnClickListener {
                        on<AlertHandler>().make().apply {
                            title = on<ResourcesHandler>().resources.getString(R.string.cancel_event)
                            message = on<ResourcesHandler>().resources.getString(R.string.event_will_be_cancelled, event!!.name)
                            positiveButton = on<ResourcesHandler>().resources.getString(R.string.cancel_event)
                            positiveButtonCallback = { result ->
                                on<DisposableHandler>().add(on<ApiHandler>().cancelEvent(event.id!!).subscribe({ successResult ->
                                    if (successResult.success) {
                                        on<DefaultAlerts>().message(on<ResourcesHandler>().resources.getString(R.string.event_cancelled, event.name))
                                        on<RefreshHandler>().refreshEvents(LatLng(event.latitude!!, event.longitude!!))
                                    } else {
                                        on<DefaultAlerts>().thatDidntWork()
                                    }
                                }, { on<DefaultAlerts>().thatDidntWork() }))
                            }
                            show()
                        }
                    }
            ))
        }

        adapter.items = items
    }

    private fun shareEvent(event: Event) {
        on<ShareActivityTransitionHandler>().shareEvent(event.id!!)
    }

    private fun isEventCancelable(event: Event?): Boolean {
        return event != null && on<PersistenceHandler>().phoneId != null &&
                !event.cancelled && event.creator != null &&
                Date().before(event.endsAt) &&
                event.creator == on<PersistenceHandler>().phoneId
    }

    private fun isEventEnded(event: Event?) = event != null && Date().after(event.endsAt)

    private fun showGroupOnMap(group: Group?) {
        (on<ActivityHandler>().activity as CircularRevealActivity)
                .finish { on<MapActivityHandler>().showGroupOnMap(group!!) }
    }

    private fun showEventOnMap(event: Event?) {
        (on<ActivityHandler>().activity as CircularRevealActivity)
                .finish { on<MapActivityHandler>().showEventOnMap(event!!) }
    }

    class ToolbarItem(
            var name: String,
            @field:DrawableRes var icon: Int,
            var onClickListener: View.OnClickListener,
            var value: ContentViewType? = null,
            @field:ColorRes var color: Int? = null,
            val indicator: BehaviorSubject<Boolean>? = null
    )
}
